.orchestra/support/distributable-test.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python3
+#
+# This file is distributed under the MIT License. See LICENSE.md for details.
+#
+
+import io
+import os
+import socket
+import sys
+import traceback
+from graphlib import TopologicalSorter
+from subprocess import Popen
+from time import sleep
+from urllib.error import URLError
+from urllib.request import urlopen
+
+from gql import Client, gql
+from gql.transport.requests import RequestsHTTPTransport
+from psutil import Process
+
+FILTER_ENV = [
+    "STARLETTE_DEBUG",
+    "REVNG_NOTIFY_FIFOS",
+    "REVNG_ORIGINS",
+    "REVNG_DATA_DIR",
+    "REVNG_PROJECT_ID",
+]
+
+
+def print_fd(fd: int):
+    os.lseek(fd, 0, io.SEEK_SET)
+    out_read = os.fdopen(fd, "r")
+    print("\n\n Daemon Stacktrace: \n")
+    print(out_read.read())
+
+
+def check_server_up(address: str):
+    for _ in range(10):
+        try:
+            req = urlopen(f"http://{address}/status", timeout=1.0)
+            if req.code == 200:
+                return
+            sleep(1.0)
+        except (URLError, TimeoutError, ConnectionResetError, socket.timeout):
+            sleep(1.0)
+    raise ValueError()
+
+
+def get_listen_port(pid: int) -> int:
+    psutil_process = Process(pid)
+    while True:
+        connections = psutil_process.connections()
+        for connection in connections:
+            if connection.raddr == () and connection.status == "LISTEN":
+                return connection.laddr.port
+        sleep(0.5)
+
+
+def main():
+    out_fd = os.memfd_create("daemon_debug", 0)
+    assert len(sys.argv) == 4
+    if sys.argv[2] == "-p":
+        process = Popen(["true"])
+        address = sys.argv[3]
+    elif sys.argv[2] == "-c":
+        out = os.fdopen(out_fd, "w")
+        process = Popen(
+            [sys.argv[3], "daemon", "-p", "0"],
+            env={k: v for k, v in os.environ.items() if k not in FILTER_ENV},
+            stdout=out,
+            stderr=out,
+            text=True,
+        )
+        port = get_listen_port(process.pid)
+        address = f"127.0.0.1:{port}"
+    else:
+        raise ValueError(f"Unknown mode: {sys.argv[2]}")
+
+    try:
+        check_server_up(address)
+    except ValueError as e:
+        print_fd(out_fd)
+        raise e
+
+    transport = RequestsHTTPTransport(f"http://{address}/graphql/")
+    client = Client(transport=transport, fetch_schema_from_transport=True)
+
+    try:
+        print("### RUNNING DISTRIBUTABLE E2E TEST ###")
+        test_revng_distributable(client)
+    except Exception as e:
+        traceback.print_exception(e.__class__, e, e.__traceback__)
+        process.communicate()
+        print_fd(out_fd)
+        sys.exit(1)
```

Comment:
Document why this is better than just the exception reach the user.

Reply:
Dropped